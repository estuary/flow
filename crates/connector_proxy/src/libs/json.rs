use doc::ptr::{Pointer, Token};
use schemars::{schema::RootSchema, JsonSchema};
use serde_json::Value;

// Create the RootSchema given datatype T.
pub fn create_root_schema<T: JsonSchema>() -> RootSchema {
    let mut settings = schemars::gen::SchemaSettings::draft07();
    settings.inline_subschemas = true;
    let generator = schemars::gen::SchemaGenerator::new(settings);
    return generator.into_root_schema_for::<T>();
}

// Extract the sub object keyed at `key` from v if v is an object, and returns both the remainder and the removed value.
pub fn remove_subobject(mut v: Value, key: &str) -> (Option<Value>, Value) {
    let mut sub_object = None;

    if let Value::Object(ref mut m) = v {
        sub_object = m.remove(key)
    }

    (sub_object, v)
}

pub fn tokenize_jsonpointer(ptr: &str) -> Vec<String> {
    Pointer::from_str(&ptr)
        .iter()
        .map(|t| match t {
            // Keep the index and next index for now. Could adjust based on usecases.
            Token::Index(ind) => ind.to_string(),
            Token::Property(prop) => prop.to_string(),
            Token::NextIndex => "-".to_string(),
        })
        .collect()
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn test_tokenize_jsonpointer() {
        let expected: Vec<String> = vec!["p1", "p2", "56", "p3", "-"]
            .iter()
            .map(|s| s.to_string())
            .collect();

        assert!(expected == tokenize_jsonpointer("/p1/p2/56/p3/-"));
    }
}
