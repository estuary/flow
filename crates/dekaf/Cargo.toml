
[package]
name = "dekaf"
version.workspace = true
rust-version.workspace = true
edition.workspace = true
authors.workspace = true
homepage.workspace = true
repository.workspace = true
license.workspace = true

[dependencies]
avro = { path = "../avro" }
doc = { path = "../doc" }
gazette = { path = "../gazette" }
labels = { path = "../labels" }
ops = { path = "../ops" }
flowctl = { path = "../flowctl" }
proto-flow = { path = "../proto-flow" }
proto-gazette = { path = "../proto-gazette" }
simd-doc = { path = "../simd-doc" }

anyhow = { workspace = true }
axum = { workspace = true }
axum-extra = { workspace = true }
base64 = { workspace = true }
bytes = { workspace = true }
clap = { workspace = true }
futures = { workspace = true }
itertools = { workspace = true }
kafka-protocol = { workspace = true }
md5 = { workspace = true }
postgrest = { workspace = true }
rsasl = { workspace = true }
serde = { workspace = true }
serde_json = { workspace = true }
socket2 = { workspace = true }
time = { workspace = true }
tokio = { workspace = true }
tokio-util = { workspace = true }
tokio-rustls = { workspace = true }
tower-http = { workspace = true }
tracing = { workspace = true }
tracing-subscriber = { workspace = true }
typestate = { workspace = true }
url = { workspace = true }
rustls-native-certs = { workspace = true }
deadpool = { workspace = true }
regex = { workspace = true }
hex = { workspace = true }
crypto-common = { workspace = true }
aes-siv = { workspace = true }
percent-encoding = { workspace = true }
rustls = { workspace = true }
metrics = { workspace = true }
metrics-prometheus = { workspace = true }
prometheus = { workspace = true }
hexdump = { workspace = true }
