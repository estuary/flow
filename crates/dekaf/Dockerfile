FROM lukemathwalker/cargo-chef:latest-rust-1 AS chef
WORKDIR /app
RUN apt-get update && apt-get install -y clang

FROM chef AS planner
COPY . .
RUN cargo chef prepare --recipe-path recipe.json

FROM chef AS builder
COPY --from=planner /app/recipe.json recipe.json
# Build dependencies - this is the caching Docker layer!
RUN cargo chef cook --release --recipe-path recipe.json
# Build application
COPY . .
RUN cargo build --release --bin dekaf

# We do not need the Rust toolchain to run the binary!
FROM debian:bookworm-slim AS runtime
WORKDIR /app
COPY --from=builder /app/target/release/dekaf /
RUN --mount=type=cache,target=/var/cache/apt \
    --mount=type=cache,target=/var/lib/apt \
    apt-get update && apt-get install -y libssl3 ca-certificates
CMD ["/dekaf"]