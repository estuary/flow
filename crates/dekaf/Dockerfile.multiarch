# Multi-architecture Dockerfile for Dekaf
# This expects pre-built binaries to be available in the build context
FROM --platform=$TARGETPLATFORM debian:trixie-slim

# Install runtime dependencies including sops and jq for config decryption
RUN --mount=type=cache,target=/var/cache/apt \
    --mount=type=cache,target=/var/lib/apt \
    apt-get update && apt-get install -y \
    libssl3 \
    ca-certificates \
    curl \
    jq \
    && rm -rf /var/lib/apt/lists/*

# Install sops (multi-arch binary from GitHub releases)
ARG TARGETPLATFORM
RUN if [ "$TARGETPLATFORM" = "linux/amd64" ]; then \
      SOPS_ARCH="amd64"; \
    elif [ "$TARGETPLATFORM" = "linux/arm64" ]; then \
      SOPS_ARCH="arm64"; \
    else \
      echo "Unsupported platform: $TARGETPLATFORM" && exit 1; \
    fi && \
    SOPS_VERSION="3.8.1" && \
    curl -L -o /usr/local/bin/sops \
      "https://github.com/getsops/sops/releases/download/v${SOPS_VERSION}/sops-v${SOPS_VERSION}.linux.${SOPS_ARCH}" && \
    chmod +x /usr/local/bin/sops

# Set up the working directory
WORKDIR /app

# Copy the appropriate binary based on the target platform
# The binaries should be in the build context at:
# - dekaf-x86_64 for amd64
# - dekaf-aarch64 for arm64
ARG TARGETPLATFORM
COPY dekaf-* /tmp/

# Select and install the correct binary for the platform
RUN if [ "$TARGETPLATFORM" = "linux/amd64" ]; then \
      mv /tmp/dekaf-x86_64 /usr/local/bin/dekaf; \
    elif [ "$TARGETPLATFORM" = "linux/arm64" ]; then \
      mv /tmp/dekaf-aarch64 /usr/local/bin/dekaf; \
    else \
      echo "Unsupported platform: $TARGETPLATFORM" && exit 1; \
    fi && \
    rm -f /tmp/dekaf-* && \
    chmod +x /usr/local/bin/dekaf

# Run as non-root user for security
RUN useradd -m -u 1000 dekaf && chown -R dekaf:dekaf /app
USER dekaf

# Default command
CMD ["dekaf"]