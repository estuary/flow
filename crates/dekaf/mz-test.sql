/*
Start local materialize as:

  docker run --network host --rm -it -v mzdata:/mzdata -p 6875:6875 -p 6876:6876 materialize/materialized

Apply this test schema as:

  psql postgres://materialize@localhost:6875/materialize --file mz-test.sql


When running using a local data-plane-gateway, use the SSL_CERT_FILE to allow
its self-signed certificate, as in:

  SSL_CERT_FILE=/home/${USER}/estuary/data-plane-gateway/local-tls-cert.pem

*/

DROP SECRET IF EXISTS estuary_token CASCADE;

CREATE SECRET estuary_token AS '...';

CREATE CONNECTION kafka_connection TO KAFKA (
    BROKER 'localhost',
    SECURITY PROTOCOL = 'SASL_PLAINTEXT',
    SASL MECHANISMS = 'PLAIN',
    SASL USERNAME = '{}',
    SASL PASSWORD = SECRET estuary_token
);

CREATE CONNECTION csr_connection TO CONFLUENT SCHEMA REGISTRY (
    URL 'http://localhost:9093',
    USERNAME = '{}',
    PASSWORD = SECRET estuary_token
);

CREATE SOURCE my_source
  FROM KAFKA CONNECTION kafka_connection (TOPIC 'demo/wikipedia/recentchange-sampled')
  FORMAT AVRO USING CONFLUENT SCHEMA REGISTRY CONNECTION csr_connection
  ENVELOPE UPSERT
  ;
