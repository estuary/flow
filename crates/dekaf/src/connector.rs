use anyhow::{bail, Context};
use proto_flow::{flow::materialization_spec, materialize};
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};
use std::collections::BTreeMap;

#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema, Copy)]
#[serde(rename_all = "snake_case")]
pub enum DeletionMode {
    // Handles deletions using the regular Kafka upsert envelope, where a deletion
    // is represented by a record containing the key that was deleted, and a null value.
    Kafka,
    // Handles deletions by passing through the full deletion document as it exists
    // in the source collection, as well as including a new field `_meta/is_deleted`
    // which is defined as the number `1` on deletions, and `0` otherwise.
    #[serde(rename = "cdc")]
    CDC,
}

impl Default for DeletionMode {
    fn default() -> Self {
        Self::Kafka
    }
}

/// Configures the behavior of a whole dekaf task
#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
pub struct DekafConfig {
    /// Whether or not to expose topic names in a strictly Kafka-compliant format
    /// for systems that require it. Off by default.
    pub strict_topic_names: bool,
    /// The password that will authenticate Kafka consumers to this task.
    // TODO(jshearer): Uncomment when schemars 1.0 is out and we upgrade
    // #[schemars(extend("secret" = true))]
    #[schemars(schema_with = "token_secret")]
    pub token: String,
    /// How to handle deletion events. "Default" emits them as regular Kafka
    /// tombstones with null values, and "Header" emits then as a kafka document
    /// with empty string and `_is_deleted` header set to `1`. Setting this value
    /// will also cause all other non-deletions to have an `_is_deleted` header of `0`.
    #[serde(default)]
    pub deletions: DeletionMode,
}

/// Configures a particular binding in a Dekaf-type materialization
#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
pub struct DekafResourceConfig {
    /// The exposed name of the topic that maps to this binding. This
    /// will be exposed through the Kafka metadata/discovery APIs.
    #[schemars(schema_with = "collection_name")]
    pub topic_name: String,
}

fn collection_name(_gen: &mut schemars::gen::SchemaGenerator) -> schemars::schema::Schema {
    serde_json::from_value(serde_json::json!({
        "x-collection-name": true,
    }))
    .unwrap()
}

fn token_secret(_gen: &mut schemars::gen::SchemaGenerator) -> schemars::schema::Schema {
    serde_json::from_value(serde_json::json!({
        "title": "Dekaf Auth Token",
        "secret": true,
    }))
    .unwrap()
}

pub async fn unary_materialize(
    request: materialize::Request,
) -> anyhow::Result<materialize::Response> {
    use proto_flow::materialize::response::validated;
    if let Some(mut validate) = request.validate {
        match materialization_spec::ConnectorType::try_from(validate.connector_type)? {
            materialization_spec::ConnectorType::Dekaf => {}
            other => bail!("invalid connector type: {}", other.as_str_name()),
        };

        let parsed_outer_config =
            serde_json::from_str::<models::DekafConfig>(&validate.config_json)
                .context("validating dekaf config")?;

        let _parsed_inner_config = serde_json::from_value::<DekafConfig>(
            parsed_outer_config.config.to_value(),
        )
        .context(format!(
            "validating dekaf endpoint config for variant {}",
            parsed_outer_config.variant
        ))?;

        // Largely copied from crates/validation/src/noop.rs
        let validated_bindings = std::mem::take(&mut validate.bindings)
            .into_iter()
            .enumerate()
            .map(|(i, b)| {
                let resource_path = vec![format!("binding-{}", i)];
                let constraints = b
                    .collection
                    .expect("collection must exist")
                    .projections
                    .into_iter()
                    .map(|proj| {
                        (
                            proj.field,
                            validated::Constraint {
                                r#type: validated::constraint::Type::FieldOptional as i32,
                                reason: "Dekaf allows everything for now".to_string(),
                            },
                        )
                    })
                    .collect::<BTreeMap<_, _>>();
                validated::Binding {
                    constraints,
                    resource_path,
                    delta_updates: false,
                }
            })
            .collect::<Vec<_>>();

        return Ok(materialize::Response {
            validated: Some(materialize::response::Validated {
                bindings: validated_bindings,
            }),
            ..Default::default()
        });
    } else {
        bail!("Unhandled request type")
    }
}
