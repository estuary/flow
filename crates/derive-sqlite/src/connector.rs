use super::{dbutil, do_validate, parse_validate, Config, Lambda, Param, Transform};
use anyhow::Context;
use futures::channel::mpsc;
use futures::TryStreamExt;
use futures::{SinkExt, Stream};
use prost::Message;
use proto_flow::runtime::{derive_response_ext, DeriveRequestExt, DeriveResponseExt};
use proto_flow::{
    derive::{request, response, Request, Response},
    flow, RuntimeCheckpoint,
};

pub fn connector<R>(
    _peek_request: &Request,
    request_rx: R,
) -> tonic::Result<impl Stream<Item = tonic::Result<Response>>>
where
    R: futures::stream::Stream<Item = tonic::Result<Request>> + Send + Unpin + 'static,
{
    let (mut response_tx, response_rx) = mpsc::channel(16);

    tokio::runtime::Handle::current().spawn_blocking(move || {
        futures::executor::block_on(async move {
            if let Err(status) = serve(request_rx, &mut response_tx).await {
                _ = response_tx.send(Err(status)).await;
            }
        })
    });

    Ok(response_rx)
}

async fn serve<R>(
    mut request_rx: R,
    response_tx: &mut mpsc::Sender<tonic::Result<Response>>,
) -> tonic::Result<()>
where
    R: futures::stream::Stream<Item = tonic::Result<Request>> + Send + Unpin + 'static,
{
    let tokio_handle = tokio::runtime::Handle::current();
    // Configured migration blocks of the last Request.Open.
    let mut migrations: Vec<String> = Vec::new();
    // Configured transform blocks and parameters of the last Request.Open.
    let mut transforms: Vec<Transform> = Vec::new();
    // A possibly opened Sqlite DB context.
    let mut maybe_handle: Option<Handle> = None;

    loop {
        match request_rx.try_next().await? {
            None => return Ok(()),
            Some(Request {
                validate: Some(validate),
                ..
            }) => {
                let validated = parse_validate(validate)
                    .and_then(|(migrations, transforms)| do_validate(&migrations, &transforms))
                    .map_err(anyhow_to_status)?;

                let _ = response_tx
                    .send(Ok(Response {
                        validated: Some(validated),
                        ..Default::default()
                    }))
                    .await;
            }
            Some(Request {
                open: Some(open),
                internal,
                ..
            }) => {
                let sqlite_uri: String;
                (sqlite_uri, migrations, transforms) =
                    parse_open(open, internal).map_err(anyhow_to_status)?;

                let (handle, runtime_checkpoint) =
                    Handle::new(&sqlite_uri, &migrations, &transforms).map_err(anyhow_to_status)?;

                // Send Opened extended with our recovered runtime checkpoint.
                let _ = response_tx
                    .send(Ok(Response {
                        opened: Some(response::Opened {}),
                        internal: Some(proto_flow::Any {
                            type_url: "flow://runtime.DeriveResponseExt".to_string(),
                            value: DeriveResponseExt {
                                opened: Some(derive_response_ext::Opened {
                                    runtime_checkpoint: Some(runtime_checkpoint),
                                }),
                                ..Default::default()
                            }
                            .encode_to_vec()
                            .into(),
                        }),

                        ..Default::default()
                    }))
                    .await;

                maybe_handle = Some(handle);
            }
            Some(Request {
                read: Some(read), ..
            }) => {
                let handle = maybe_handle
                    .as_mut()
                    .ok_or_else(|| tonic::Status::invalid_argument("Read without Open"))?;

                do_read(&mut handle.transforms, read, response_tx, &tokio_handle)
                    .map_err(anyhow_to_status)?;
            }
            Some(Request {
                flush: Some(request::Flush {}),
                ..
            }) => {
                // Send Flushed to runtime.
                let _ = response_tx
                    .send(Ok(Response {
                        flushed: Some(response::Flushed {}),
                        ..Default::default()
                    }))
                    .await;
            }
            Some(Request {
                start_commit: Some(request::StartCommit { runtime_checkpoint }),
                ..
            }) => {
                let handle = maybe_handle
                    .as_ref()
                    .ok_or_else(|| tonic::Status::invalid_argument("StartCommit without Open"))?;

                let started_commit =
                    do_commit(handle.conn, runtime_checkpoint).map_err(anyhow_to_status)?;

                // Send StartedCommit to runtime.
                let _ = response_tx
                    .send(Ok(Response {
                        started_commit: Some(started_commit),
                        ..Default::default()
                    }))
                    .await;
            }
            Some(Request {
                reset: Some(request::Reset {}),
                ..
            }) => {
                // Replace with a new :memory: database with the same configuration.
                let (db, _runtime_checkpoint) =
                    Handle::new(":memory:", &migrations, &transforms).map_err(anyhow_to_status)?;
                maybe_handle = Some(db);
            }
            Some(malformed) => {
                return Err(tonic::Status::invalid_argument(format!(
                    "invalid request {malformed:?}"
                )))
            }
        }
    }
}

fn parse_open(
    open: request::Open,
    internal: Option<proto_flow::Any>,
) -> anyhow::Result<(String, Vec<String>, Vec<Transform>)> {
    let request::Open {
        collection,
        range: _,
        state_json: _,
        version: _,
    } = open;

    let internal = internal.unwrap_or_default();
    let DeriveRequestExt { open: open_ext, .. } =
        Message::decode(internal.value).context("internal is a DeriveRequestExt")?;
    let open_ext = open_ext.unwrap_or_default();

    let mut sqlite_uri = open_ext.sqlite_vfs_uri;
    if sqlite_uri.is_empty() {
        sqlite_uri = ":memory:".to_string();
    }

    let flow::CollectionSpec { derivation, .. } = collection.unwrap();

    let flow::collection_spec::Derivation {
        config_json,
        transforms,
        ..
    } = derivation.unwrap();

    let config: Config = serde_json::from_str(&config_json)
        .with_context(|| format!("failed to parse SQLite configuration: {config_json}"))?;

    let transforms: Vec<Transform> = transforms
        .into_iter()
        .map(|transform| {
            let flow::collection_spec::derivation::Transform {
                name,
                collection: source,
                lambda_config_json,
                shuffle_lambda_config_json: _,
                ..
            } = transform;

            let source = source.unwrap();
            let params: Vec<_> = source.projections.iter().map(Param::new).collect();

            let block: String = serde_json::from_str(&lambda_config_json).with_context(|| {
                format!("failed to parse SQLite lambda block: {lambda_config_json}")
            })?;

            Ok(Transform {
                name,
                block,
                source: source.name,
                params,
            })
        })
        .collect::<Result<_, anyhow::Error>>()?;

    Ok((sqlite_uri, config.migrations, transforms))
}

fn do_read<'db>(
    transforms: &mut [Vec<Lambda<'db>>],
    read: request::Read,
    response_tx: &mut mpsc::Sender<tonic::Result<Response>>,
    tokio_handle: &tokio::runtime::Handle,
) -> anyhow::Result<()> {
    let request::Read {
        transform,
        doc_json,
        uuid: _,
        shuffle: _,
    } = read;

    let stack = transforms
        .get_mut(transform as usize)
        .with_context(|| format!("invalid transform index {transform}"))?;

    let doc: serde_json::Value = serde_json::from_str(&doc_json)
        .with_context(|| format!("couldn't parse read document as JSON: {doc_json}",))?;

    // Invoke each lambda of the stack in turn, streaming published documents into `response_tx`.
    // It's important that we don't block here -- these result sets could be very large.
    for (index, lambda) in stack.iter_mut().enumerate() {
        let it = lambda
            .invoke(&doc)
            .with_context(|| format!("failed to invoke lambda statement at offset {index}"))?;

        let it = it.map(|published| match published {
            Ok(published) => Ok(Ok(Response {
                published: Some(response::Published {
                    doc_json: published.to_string(),
                }),
                ..Default::default()
            })),
            Err(err) => Ok(Err(anyhow_to_status(err.into()))),
        });

        _ = tokio_handle.block_on(response_tx.send_all(&mut futures::stream::iter(it)));
    }
    Ok(())
}

fn do_commit(
    conn: &rusqlite::Connection,
    runtime_checkpoint: Option<RuntimeCheckpoint>,
) -> anyhow::Result<response::StartedCommit> {
    if let Some(runtime_checkpoint) = runtime_checkpoint {
        let () = dbutil::update_checkpoint(conn, runtime_checkpoint)?;
    }
    dbutil::commit_and_begin(conn)?;

    Ok(response::StartedCommit { state: None })
}

struct Handle {
    conn: &'static rusqlite::Connection,
    transforms: Vec<Vec<Lambda<'static>>>,
}

impl Handle {
    fn new(
        sqlite_uri: &str,
        migrations: &[String],
        transforms: &[Transform],
    ) -> anyhow::Result<(Handle, RuntimeCheckpoint)> {
        let (conn, runtime_checkpoint) = dbutil::open(sqlite_uri, &migrations)?;

        // Place into Self so it's covered by our Drop implementation.
        let mut db = Self {
            conn: Box::leak(Box::new(conn)),
            transforms: Vec::new(),
        };
        db.transforms = dbutil::build_transforms(&db.conn, &transforms)?;

        Ok((db, runtime_checkpoint))
    }
}

impl Drop for Handle {
    fn drop(&mut self) {
        // Force lambdas to drop first, before the database Connection.
        self.transforms.clear();

        // Take ownership of the boxed Connection to drop it.
        let db: *const _ = self.conn as *const _;
        let db: *mut _ = db as *mut rusqlite::Connection;
        unsafe { Box::from_raw(db) };
    }
}

fn anyhow_to_status(err: anyhow::Error) -> tonic::Status {
    tonic::Status::internal(format!("{err:?}"))
}
