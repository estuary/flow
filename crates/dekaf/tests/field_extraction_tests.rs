use anyhow::Context;
use dekaf::connector::{DekafConfig, DeletionMode};
use itertools::Itertools;
use serde_json::json;

async fn get_extraction_components(
    fixture_path: String,
) -> anyhow::Result<(
    doc::Shape,
    DekafConfig,
    proto_flow::flow::FieldSelection,
    Vec<proto_flow::flow::Projection>,
)> {
    let source = build::arg_source_to_url(fixture_path.as_ref(), false)
        .context("Failed to create source URL")?;

    let build::Output { built, .. } = managed_build(source).await;

    if !built.errors.is_empty() {
        anyhow::bail!("Build errors: {:?}", built.errors);
    }

    let materialization = built
        .built_materializations
        .get_key(&models::Materialization::new("test/materialization"))
        .context(format!(
            "Couldn't find test materialization. Found: {}",
            built
                .built_materializations
                .iter()
                .map(|mat| mat.materialization.clone())
                .join(", ")
        ))?;

    let binding = materialization
        .spec
        .as_ref()
        .expect("Missing spec")
        .bindings
        .iter()
        .next()
        .context("Materialization has no bindings")?;

    let collection_spec = built
        .built_collections
        .get_key(&models::Collection::new(
            &binding
                .collection
                .as_ref()
                .expect("missing binding collection")
                .name,
        ))
        .context("Couldn't find bound collection")?
        .spec
        .as_ref()
        .context("Collection spec is missing")?;

    let schema = if collection_spec.read_schema_json.len() > 0 {
        &collection_spec.read_schema_json
    } else {
        &collection_spec.write_schema_json
    };
    let shape = json_schema_to_shape(schema)?;

    let endpoint_config =
        dekaf::extract_dekaf_config(&materialization.spec.clone().expect("missing spec"))
            .await
            .expect("Failed to extract DekafConfig");

    Ok((
        shape,
        endpoint_config,
        binding
            .field_selection
            .clone()
            .expect("missing field selection"),
        collection_spec.projections.clone(),
    ))
}

fn json_schema_to_shape(schema: &str) -> anyhow::Result<doc::Shape> {
    let json_schema = doc::validation::build_bundle(schema)?;
    let validator = doc::Validator::new(json_schema)?;
    Ok(doc::Shape::infer(
        &validator.schemas()[0],
        validator.schema_index(),
    ))
}

async fn managed_build(source: url::Url) -> build::Output {
    use tables::CatalogResolver;

    let file_root = std::path::Path::new("/");
    let draft = build::load(&source, file_root).await;
    if !draft.errors.is_empty() {
        return build::Output::new(draft, Default::default(), Default::default());
    }
    let catalog_names = draft.all_spec_names().collect();
    let live = build::NoOpCatalogResolver.resolve(catalog_names).await;
    if !live.errors.is_empty() {
        return build::Output::new(draft, live, Default::default());
    }

    build::validate(
        models::Id::new([32; 8]),
        models::Id::new([1; 8]),
        true,
        "",
        ops::tracing_log_handler,
        false,
        false,
        false,
        &build::project_root(&source),
        draft,
        live,
    )
    .await
}

async fn roundtrip(
    fixture_path: String,
    docs: Vec<serde_json::Value>,
) -> anyhow::Result<Vec<Result<apache_avro::types::Value, apache_avro::Error>>> {
    let (shape, endpoint_config, field_selection, projections) =
        get_extraction_components(fixture_path).await?;

    let (avro_schema, extractors) = dekaf::utils::build_field_extractors(
        shape,
        field_selection,
        projections,
        endpoint_config.deletions,
    )?;

    extract_and_decode(docs, extractors, avro_schema)
}

fn extract_and_decode(
    docs: Vec<serde_json::Value>,
    extractors: Vec<(apache_avro::Schema, dekaf::utils::CustomizableExtractor)>,
    avro_schema: apache_avro::Schema,
) -> anyhow::Result<Vec<Result<apache_avro::types::Value, apache_avro::Error>>> {
    docs.into_iter()
        .map(|doc| {
            let mut encoded = Vec::new();
            dekaf::extract_and_encode(extractors.as_slice(), &doc, &mut encoded)?;

            Ok(apache_avro::from_avro_datum(
                &avro_schema,
                &mut encoded.as_slice(),
                None,
            ))
        })
        .collect::<Result<Vec<_>, _>>()
}

#[tokio::test]
async fn test_allof_with_null_default() -> anyhow::Result<()> {
    let fixture_path = "tests/fixtures/allof_with_null_default.yaml".to_string();
    let docs = vec![json!({
        "id": 1234
    })];

    for output in roundtrip(fixture_path, docs).await? {
        insta::assert_debug_snapshot!(output?);
    }

    Ok(())
}

#[tokio::test]
async fn test_field_selection_specific_fields() -> anyhow::Result<()> {
    let fixture_path = "tests/fixtures/field_selection_specific_fields.yaml".to_string();
    let docs = vec![json!({
        "key": "first",
        "field_a": "foo",
        "field_b": "bar"
    })];

    for output in roundtrip(fixture_path, docs).await? {
        insta::assert_debug_snapshot!(output?);
    }

    Ok(())
}

#[tokio::test]
async fn test_field_selection_recommended_fields() -> anyhow::Result<()> {
    let fixture_path = "tests/fixtures/field_selection_recommended_fields.yaml".to_string();
    let docs = vec![json!({
        "key": "first",
        "field_a": "foo",
        "field_b": "bar"
    })];

    for output in roundtrip(fixture_path, docs).await? {
        insta::assert_debug_snapshot!(output?);
    }

    Ok(())
}

#[tokio::test]
async fn test_deletions() -> anyhow::Result<()> {
    let fixture_path = "tests/fixtures/deletions.yaml".to_string();
    let docs = vec![
        json!({
            "key": "first",
            "_meta": { "op": "c" }
        }),
        json!({
            "key": "second",
            "_meta": { "op": "d" }
        }),
    ];

    for (idx, output) in roundtrip(fixture_path, docs).await?.into_iter().enumerate() {
        insta::assert_debug_snapshot!(format!("deletions-{}", idx), output?);
    }

    Ok(())
}

#[tokio::test]
async fn test_old_style_deletions() -> anyhow::Result<()> {
    let shape = json_schema_to_shape(
        r#"{
       "properties": {
            "key": {
                "type": "string"
            },
            "_meta": {
                "properties": {
                    "op": {
                        "type": "string"
                    }
                },
                "type": "object"
            }
        },
        "additionalProperties": {
            "type": "string"
        },
        "type": "object",
        "required": [
            "key",
            "_meta"
        ]
    }"#,
    )?;

    let (avro_schema, extractors) =
        dekaf::utils::build_LEGACY_field_extractors(shape, DeletionMode::CDC)?;

    let decoded = extract_and_decode(
        vec![
            json!({
                "key": "first",
                "_meta": {
                    "op": "c"
                },
            }),
            json!({
                "key": "second",
                "_meta": {
                    "op": "d"
                },
            }),
            json!({
                "key": "second",
                "_meta": {
                    "op": "d"
                },
                "additional": "I should end up in _flow_extra"
            }),
        ],
        extractors,
        avro_schema,
    )?;

    for (idx, doc) in decoded.into_iter().enumerate() {
        insta::assert_debug_snapshot!(format!("old-deletions-{}", idx), doc?);
    }

    Ok(())
}
