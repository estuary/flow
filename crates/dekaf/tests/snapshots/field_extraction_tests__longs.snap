---
source: crates/dekaf/tests/field_extraction_tests.rs
expression: "roundtrip(fixture_path.clone(), raw_input.as_bytes()).await"
---
Ok(
    [
        Ok(
            Record(
                [
                    (
                        "key",
                        String(
                            "1234",
                        ),
                    ),
                    (
                        "flow_published_at",
                        Union(
                            0,
                            Record(
                                [
                                    (
                                        "json",
                                        String(
                                            "null",
                                        ),
                                    ),
                                ],
                            ),
                        ),
                    ),
                    (
                        "my_long",
                        Union(
                            0,
                            Long(
                                0,
                            ),
                        ),
                    ),
                ],
            ),
        ),
        Ok(
            Record(
                [
                    (
                        "key",
                        String(
                            "1234",
                        ),
                    ),
                    (
                        "flow_published_at",
                        Union(
                            0,
                            Record(
                                [
                                    (
                                        "json",
                                        String(
                                            "null",
                                        ),
                                    ),
                                ],
                            ),
                        ),
                    ),
                    (
                        "my_long",
                        Union(
                            0,
                            Long(
                                10,
                            ),
                        ),
                    ),
                ],
            ),
        ),
    ],
)
